File: hw2/hw1/Parser.cs
using System;
using System.Collections;
using System.Linq;

namespace hw1
{
    public static class Parser
    {
        private static readonly string[] ExpectedOperation = new[]
        {
            "+",
            "-",
            "*",
            $"/"
        };

        public static int TryToParse(string[] args, out int val1, out string operation, out int val2)
        {
            var isVal1Int = int.TryParse(args[0], out val1);
            operation = args[1];
            var isVal2Int = int.TryParse(args[2], out val2);

            if (!isVal1Int || !isVal2Int) //if there are no int args
            {
                Console.WriteLine($"{args[0]}{args[1]}{args[2]} is not a valid calculation syntax");
                return 1;
            }

            if (((IList) ExpectedOperation).Contains(operation)) return 0;
            
            Console.WriteLine(
                $"{args[0]}{args[1]}{args[2]} is not a valid calculation syntax. "
                + $"Supported operations are "
                + $"{ExpectedOperation.Aggregate((c, n) => $"{c} {n}")}");
            return 2;//if unsupported operation

        }
    }
}

File: hw4/hw4.FS/Parser.fs
module hw4.FS.Parser

    open System
    let ExpectedOperation =
        [
            "+";
            "-";
            "*";
            "/"
        ]
        
    let TryToParse (args:string[]) (val1 : outref<int>) (operation : outref<string> ) (val2 : outref<int>) =
        let tryParseWith (tryParseFunc: string -> bool * _) = tryParseFunc >> function
            | true, v    -> Some v
            | false, _   -> None
        let parseInt = tryParseWith Int32.TryParse
        let exists (x : int option) =
            match x with
            | Some(x) -> true
            | None -> false
            
        if args.Length <> 3 then
            printfn $"3 arguments were expected ! Not {args.Length}"
            1
        elif List.contains args.[1] ExpectedOperation = false then 2 //if unsupported operation
        elif parseInt args.[0] |> exists = false || parseInt args.[2] |> exists = false
                || (args.[1] = "/" && args.[2] = "0")
                 then 1 //if there are no int args or DivideByZero
        else
            val1 <- args.[0] |> int
            operation <- args.[1]
            val2 <- args.[2] |> int
            0